A29-xds-tls-security.md

```diff
@@ -0,0 +1,411 @@
+of the matched (selected) `filter_chain` is used to extract the
+[`DownstreamTlsContext`](https://github.com/envoyproxy/envoy/blob/main/api/envoy/extensions/transport_sockets/tls/v3/tls.proto#L57)
+as described
+[here](https://github.com/envoyproxy/envoy/blob/main/api/envoy/config/listener/v3/listener_components.proto#L246).
+The `DownstreamTlsContext` thus obtained is used for the incoming connection.
+
+[`common_tls_context`](https://github.com/envoyproxy/envoy/blob/main/api/envoy/extensions/transport_sockets/tls/v3/tls.proto#L84)
+in the `DownstreamTlsContext` contains the required security configuration. The
+[`require_client_certificate`](https://github.com/envoyproxy/envoy/blob/main/api/envoy/extensions/transport_sockets/tls/v3/tls.proto#L88) field
+determines whether the client certificate is required (mTLS mode if `true`, TLS if `false`).
+All the other fields are ignored. See below for
+[`CommonTlsContext`](https://github.com/envoyproxy/envoy/blob/main/api/envoy/extensions/transport_sockets/tls/v3/tls.proto#L129)
+processing.
+
+
+#### CommonTlsContext Processing
+
+`CommonTlsContext` is present in both the `UpstreamTlsContext` and `DownstreamTlsContext` protos and contains the
+certificate and key configuration information needed on the client and server side respectively. The configuration
+tells gRPC how to obtain certificates and the keys for the TLS handshake. Although there are various ways to obtain
+certificates as per this proto (which are supported by Envoy), gRPC supports only one of them and that is defined by the
+[`CertificateProviderInstance`](https://github.com/envoyproxy/envoy/blob/main/api/envoy/extensions/transport_sockets/tls/v3/tls.proto#L154)
+proto which is based on the notion of CertificateProvider plugin framework (described later).
+The field `instance_name` defines a CertificateProvider "instance" that gRPC looks up in the
+bootstrap file (described later) to obtain the CertificateProvider configuration. This
+configuration along with the CertificateProvider plugin framework enables gRPC to acquire
+the certificates necessary for the TLS handshake.
+
+The field [`tls_certificate_certificate_provider_instance`](https://github.com/envoyproxy/envoy/blob/main/api/envoy/extensions/transport_sockets/tls/v3/tls.proto#L234)
+is used for the local certificate and the private key. And either
+[`validation_context_certificate_provider_instance`](https://github.com/envoyproxy/envoy/blob/main/api/envoy/extensions/transport_sockets/tls/v3/tls.proto#L259)
+or [`validation_context_certificate_provider_instance`](https://github.com/envoyproxy/envoy/blob/main/api/envoy/extensions/transport_sockets/tls/v3/tls.proto#L200)
+inside [`combined_validation_context`](https://github.com/envoyproxy/envoy/blob/main/api/envoy/extensions/transport_sockets/tls/v3/tls.proto#L251)
+is used for the root certificate-chain for validating peer certificates. For mTLS both the local certificate
```

Comment:
We should probably also mention the behavior if the instance_name is not recognized, i.e., not present in the bootstrap file. 
Core's behavior - Don't check at parsing time, so no NACKing. It's checked later when it needs to be used, at which point an error is propagated. I believe that results in a TransientFailurePicker on the client side and closing the connection on the server side.

Reply:
gRPC-Go NACKs the resource if `require_client_certificate` is set and a certificate provider instance name for the root certificate is not provided. I think we discussed this long back and decided to NACK in this case.

Reply:
> We should probably also mention the behavior if the instance_name is not recognized, i.e., not present in the bootstrap file.

In gRPC-Go, we do something similar to Core. In fact, we cannot verify at parsing time as to whether the given certificate provider instance name is present in the bootstrap config or not. So, we simply accept it. But when the update it given to the CDS balancer, it is able to verify if the instance name is present in the bootstrap config or not. If the instance name is not found in the bootstrap config, the CDS balancer propagates such errors down to its child policy, which eventually leads to the channel being put in transient failure.

On the server side, since we are not required to put the server in `not-serving` mode for bad configuration (we are only required to do that when the configuration is deleted), we simply fail the connection at handshake time if the provider instance is not found in the bootstrap file.

Reply:
Why can't we validate that the certificate provider instance is valid at config parsing time?  I think that would be a trivial change in C-core, and it seems like the right thing to do.  We basically do the same thing for supported HTTP filters (the only difference being that those are not configured in the bootstrap file, they're set at compile time).

Reply:
> Why can't we validate that the certificate provider instance is valid at config parsing time?

That's how the code is modularized. The component which handles receiving the message and parsing knows only to do that. It does not have knowledge of the contents of the bootstrap file. I don't think it will super hard to give it that information though. But I think we discussed this in some form when we were implementing, and decided not to NACK, but instead to fail things somewhere up the stack where the required information is already available.

I don't remember the exact the pros and cons of the argument though. @yashykt @sanjaypujare

Reply:
Yes, checking the instance name is possible for Core, but I think we decided to go with <current_behavior> because of limitations from Go and to avoid inconsistencies between languages.

Reply:
If it's not super hard to do this in any language (and I don't see why it would be; it's just passing in a list of supported instance names to the config validation code), then I think we should do this at config-validation time.  We should NACK if we are told to use a provider that is not configured.

Reply:
Sanjay, please clarify in the text that we will NACK if the specified certificate provider instance name is not one of the instances configured in the bootstrap file.